File: src/types/team-template.ts
/**
 * @module types/team-template
 * @description TypeScript type definitions for team templates, orchestration patterns, and team library
 *
 * PURPOSE:
 * - Define TeamTemplate interface for saved team configurations
 * - Define LibraryTeamTemplate for pre-built team catalog
 * - Define OrchestrationPattern for team coordination styles
 * - Define teammate, task, and hook sub-types
 *
 * EXPORTS:
 * - OrchestrationPattern - "leader" | "pipeline" | "parallel" | "swarm" | "council"
 * - TeamCategory - Categories for organizing teams in the library
 * - TeammateDef - Definition of a single teammate in a team
 * - TeamTaskDef - Definition of a task with dependencies
 * - TeamHookDef - Definition of a hook for team coordination
 * - LibraryTeamTemplate - A pre-defined team from the library catalog
 * - TeamTemplate - A saved team template with database fields
 * - ScoredTeamTemplate - A LibraryTeamTemplate with relevance score
 * - TeamCategoryInfo - Metadata about a team category (label, description, icon)
 * - ProjectContext - Active project tech stack context for personalizing deploy output
 *
 * PATTERNS:
 * - Types mirror Rust structs in models/team_template.rs
 * - Team templates have usage analytics (usageCount)
 * - LibraryTeamTemplate has tags for tech-stack-based relevance scoring
 *
 * CLAUDE NOTES:
 * - Keep in sync with Rust models in src-tauri/src/models/team_template.rs
 * - DateTime fields are serialized as ISO strings by Tauri
 * - TechTag is imported from skill.ts for consistency
 * - Orchestration patterns define how teammates coordinate
 */

import type { TechTag } from "./skill";

export type OrchestrationPattern = "leader" | "pipeline" | "parallel" | "swarm" | "council";

export type TeamCategory =
  | "feature-development"
  | "testing"
  | "code-review"
  | "refactoring"
  | "migration"
  | "documentation"
  | "devops";

export interface TeammateDef {
  role: string;
  description: string;
  spawnPrompt: string;
}

export interface TeamTaskDef {
  id: string;
  title: string;
  description: string;
  assignedTo: string;
  blockedBy: string[];
}

export interface TeamHookDef {
  event: string;
  command: string;
  description: string;
}

export interface LibraryTeamTemplate {
  slug: string;
  name: string;
  description: string;
  orchestrationPattern: OrchestrationPattern;
  category: TeamCategory;
  tags: TechTag[];
  teammates: TeammateDef[];
  tasks: TeamTaskDef[];
  hooks: TeamHookDef[];
  leadSpawnInstructions: string;
}

export interface TeamTemplate {
  id: string;
  name: string;
  description: string;
  orchestrationPattern: OrchestrationPattern;
  category: TeamCategory;
  teammates: TeammateDef[];
  tasks: TeamTaskDef[];
  hooks: TeamHookDef[];
  leadSpawnInstructions: string;
  projectId: string | null;
  usageCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface ScoredTeamTemplate {
  template: LibraryTeamTemplate;
  score: number;
  isRecommended: boolean;
  matchedTags: TechTag[];
}

export interface TeamCategoryInfo {
  id: TeamCategory;
  label: string;
  description: string;
  icon: string;
}

export interface ProjectContext {
  name: string;
  language: string | null;
  framework: string | null;
  testFramework: string | null;
  buildTool: string | null;
  styling: string | null;
  database: string | null;
}

File: src-tauri/src/models/team_template.rs
//! @module models/team_template
//! @description Data models for team templates and orchestration patterns
//!
//! PURPOSE:
//! - Define TeamTemplate struct for saved team configurations
//! - Define TeammateDef, TeamTaskDef, TeamHookDef sub-types
//!
//! DEPENDENCIES:
//! - serde - Serialization for Tauri IPC
//! - chrono - Timestamp handling
//!
//! EXPORTS:
//! - TeamTemplate - A saved team template configuration
//! - TeammateDef - Definition of a single teammate
//! - TeamTaskDef - Definition of a task with dependencies
//! - TeamHookDef - Definition of a hook for coordination
//! - ProjectContext - Active project tech stack context for deploy output personalization
//!
//! PATTERNS:
//! - Team templates have JSON-serialized teammates, tasks, hooks
//! - Pattern field is a string enum (leader/pipeline/parallel/swarm/council)
//!
//! CLAUDE NOTES:
//! - Keep in sync with TypeScript types in src/types/team-template.ts
//! - teammates, tasks, hooks are stored as JSON text in SQLite

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TeammateDef {
    pub role: String,
    pub description: String,
    pub spawn_prompt: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TeamTaskDef {
    pub id: String,
    pub title: String,
    pub description: String,
    pub assigned_to: String,
    pub blocked_by: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TeamHookDef {
    pub event: String,
    pub command: String,
    pub description: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectContext {
    pub name: Option<String>,
    pub language: Option<String>,
    pub framework: Option<String>,
    pub test_framework: Option<String>,
    pub build_tool: Option<String>,
    pub styling: Option<String>,
    pub database: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TeamTemplate {
    pub id: String,
    pub name: String,
    pub description: String,
    pub orchestration_pattern: String,
    pub category: String,
    pub teammates: Vec<TeammateDef>,
    pub tasks: Vec<TeamTaskDef>,
    pub hooks: Vec<TeamHookDef>,
    pub lead_spawn_instructions: String,
    pub project_id: Option<String>,
    pub usage_count: u32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}
